[Mobile] Link picker via fetch

mkevins opened the pull request:
### Related PRs

`gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2484
`WordPress-Android`: https://github.com/wordpress-mobile/WordPress-Android/pull/12438
`WordPress-iOS`: https://github.com/wordpress-mobile/WordPress-iOS/pull/14537

### Details

This is a PR to produce a test apk for design review of the link picker feature.

To test:
Follow testing steps here: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2484#issue-448736366


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

mkevins committed:
Implement link cell component for RN link picker experiment

mkevins committed:
Use link cell in link modal UI

mkevins committed:
Add link cell component

mkevins committed:
Add link suggestion item cell component

mkevins committed:
Implement link picker results component via fetch function

mkevins committed:
Add link picker component

mkevins committed:
Use link picker as subsheet in format library link modal

mkevins committed:
Merge branch 'master' into rnmobile/feature/link-picker-via-fetch

mkevins committed:
Support search API endpoint

mkevins committed:
Use ListHeaderComponent HACK for nested flatlist in scrollview

This should not be necessary, and there should be a better way to
achieve the performance benefits from a flatlist nested with the same
orientation of an ancestor, especially when the scroll dimension can be
fixed.

mkevins committed:
Merge branch 'master' into rnmobile/feature/link-picker-via-fetch

mkevins committed:
Revert "Use ListHeaderComponent HACK for nested flatlist in scrollview"

This reverts commit c8adbb8a44087038ed84ffbd69fbe3ad06a259dd.

dratwas committed:
use optimised FlatList inside bottom-sheet

mkevins committed:
Merge branch 'master' into rnmobile/feature/link-picker-via-fetch-without-navigation

mkevins committed:
Merge remote-tracking branch 'origin/rnmobile/bottom-sheet-flat-list' into rnmobile/feature/link-picker-via-fetch-without-navigation

mkevins committed:
Implement basic pagination

mkevins committed:
Merge branch 'master' into rnmobile/feature/link-picker-via-fetch-without-navigation

dratwas committed:
fix android scroll on separator item

mkevins committed:
Merge branch 'rnmobile/bottom-sheet-flat-list' into rnmobile/feature/link-picker-via-fetch-without-navigation

mkevins committed:
Fix prettier issues

mkevins committed:
Debounce fetch on query changes

mkevins committed:
Add gridicons as one-offs

mkevins committed:
Merge branch 'master' into rnmobile/feature/link-picker-via-fetch-without-navigation

geriux reviewed (commented):
This is looking great already!

I know the design review is still on going but just wanted to mention this just in case it hasn't come up yet:

- I think we should add a loader somewhere to show data is being requested. (For the suggestions)
- On iOS we need to add the `SafeAreaView`

mkevins committed:
Use default instead of named import of styles in link suggestion item

mkevins committed:
Use hook instead of hoc in suggestion item

mkevins committed:
Use array literal for style instead of compose

Since we are always combining these styles, there is no need to preserve
referential equality.

dratwas committed:
fix bottom spacing in inserter and justifyContent

dratwas committed:
add crollableContentStyle and use it

dratwas committed:
remove unused style

dratwas committed:
render spacing view only if needed

mkevins committed:
Add refresh gridicon

mkevins committed:
Implement animated loading spinner

mkevins committed:
Add loading icon to link picker results

mkevins committed:
Merge branch 'rnmobile/bottom-sheet-flat-list' into rnmobile/feature/link-picker-via-fetch-without-navigation

mkevins committed:
Add temporary padding for testing apk

This should not be necessary, but the view does not scroll all the way
to the bottom without it, and the testing apk needs to reveal the
loading spinner for design feedback

mkevins commented on packages/components/src/mobile/bottom-sheet/link-suggestion-item-cell.native.js:
```diff
@@ -0,0 +1,89 @@
+	suggestionTitleStyle,
+	suggestionSummaryStyle,
+	suggestionContainerStyle,
+	hidden,
```
I don't know if we've ever used named imports for scss. I changed it to import them as default. :+1:

mkevins commented on packages/components/src/mobile/bottom-sheet/link-suggestion-item-cell.native.js:
```diff
@@ -0,0 +1,89 @@
+function LinkSuggestionItemCell( {
+	suggestion,
+	onLinkPicked,
+	getStylesFromColorScheme,
```
Thanks for the suggestion. Updated.

mkevins commented on packages/components/src/mobile/bottom-sheet/link-suggestion-item-cell.native.js:
```diff
@@ -0,0 +1,89 @@
+	);
+
+	const titleStyle = compose( cellTitleStyle, suggestionTitleStyle );
+	const summaryStyle = compose( cellSummaryStyle, suggestionSummaryStyle );
```
The difference is that `compose` preserves referential equality, where possible, reducing the number of re-renders of child components ([more details](https://reactnative.dev/docs/stylesheet#compose)). In our case, an array literal is ok since we'll _always_ need to combine the styles, so that point is moot. I've updated to use the array. :+1:

dratwas committed:
Update packages/components/src/mobile/bottom-sheet/index.native.js

Co-authored-by: Luke Walczak <lukasz.walczak.pwr@gmail.com>

dratwas committed:
fix syntax error after applying suggestion

dratwas committed:
fix: set correct height of container

mkevins committed:
Align link cell and link picker labels

mkevins committed:
Add chevron to link cell

mkevins committed:
Left align query input and remove icon

mkevins committed:
Use native spinner with some padding

mkevins committed:
Add apply button to link picker fullsheet

mkevins committed:
Refactor picker spinner to use hasAllSuggestions state

mkevins committed:
Merge branch 'rnmobile/bottom-sheet-flat-list' into rnmobile/feature/link-picker-via-fetch-without-navigation

mkevins committed:
Use apply instead of check icon in ios navigation header

mkevins committed:
Use safearea view for fullsheet link picker

mkevins commented:
Thanks for the review Gerardo! I made the suggested changes, and also addressed the following points:

> * I think we should add a loader somewhere to show data is being requested. (For the suggestions)
> * On iOS we need to add the `SafeAreaView`

enejb commented on packages/components/src/mobile/gridicons/index.native.js:
```diff
@@ -0,0 +1,25 @@
+import { SVG, Path } from '@wordpress/primitives';
+
+export const posts = (
+	<SVG xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
```
Since `<SVG xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">` is always the same should create a wrapper component for this? 

something like 
```
function svg( component ) {
   return <SVG xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">{component}</SVG>
}
```

and then we could have 
```
export const posts = svg( <Path d="M16 19H3v-2h13v2zm5-10H3v2h18V9zM3 5v2h11V5H3zm14 0v2h4V5h-4zm-6 8v2h10v-2H11zm-8 0v2h5v-2H3z" /> );
```

mkevins replied:
Thanks for the suggestion. I've updated it to reduce the duplication. Down the road, depending on how many icons we port, it may be worth it to consider something like [react-native-svg-transformer](https://github.com/kristerkari/react-native-svg-transformer).

antonis commented on packages/components/src/mobile/link-picker/index.native.js:
```diff
@@ -0,0 +1,93 @@
+
+	if ( protocol.includes( 'tel' ) ) {
+		type = 'tel';
+	}
```
I'm not sure if there is such case but the above checks can be enhanced to be case insensitive (e.g. the link may be MAILTO or Mailto)

mkevins replied:
Good idea Antonis! I've updated to make the protocol detection case-insensitive. :+1:

antonis reviewed (approved):
This LGTM @mkevins! Great job 🎉 
Tested on iOS and Android,
I just left one minor comment/suggestion.

mkevins committed:
Make protocol detection case insensitive

mkevins committed:
DRY up gridicons code a bit

mkevins committed:
Merge branch 'master' into rnmobile/feature/link-picker-via-fetch-without-navigation

mkevins committed:
Only render cell label if it is present

mkevins committed:
Adjust styles for picker input cell

mkevins committed:
Fix css lint

mkevins committed:
Add scheme-specific suggestion summaries

mkevins committed:
Use "Close" for iOS back button in fullsheet

mkevins committed:
Make link suggestion summary text smaller

mkevins committed:
Use px unit in link suggestion styles

mkevins committed:
Merge branch 'master' into rnmobile/feature/link-picker-via-fetch-without-navigation

dratwas committed:
Merge branch 'master' into rnmobile/feature/link-picker-via-fetch-without-navigation

dratwas committed:
[WIP]Refactor to use new navigation

dratwas committed:
change modalUI to function

dratwas committed:
first version of working full height modal with keyboard

dratwas committed:
[WIP] generic way of handling fullscreen bottom-sheet screens

dratwas committed:
add fullHeight prop to navigation-screen

dratwas committed:
Add setHeight to useCallback dependencies

dratwas committed:
Merge branch 'rnmobile/feat/full-height-bottom-sheet-screen' into rnmobile/feature/link-picker-via-fetch-without-navigation

dratwas committed:
Fix issue with empty records in link picker list

dratwas committed:
fix some performance issues

dratwas committed:
remove useCallbacks since it is slower and fix the focus loop

dratwas committed:
fix padding bottom

dratwas committed:
Fix for bottom insets

dratwas committed:
move styles to scss

dratwas committed:
add missing spacing on the top

dratwas committed:
fix issue with wrong title when removing the old one

dratwas committed:
Refactor: move screens to separate file

dratwas committed:
Apply suggestions from design review

dratwas committed:
remove top handler from full screen bottom sheet android

dratwas committed:
Merge branch 'master' into rnmobile/feature/link-picker-via-fetch-without-navigation

dratwas committed:
fix import of styles

dratwas committed:
fix rest of imports

dratwas committed:
Merge branch 'master' into rnmobile/feature/link-picker-via-fetch-without-navigation

dratwas committed:
Add page param to docs

dratwas committed:
set type of suggestions to post

dratwas committed:
add cancel-circle-filled icon and use it in link picker

dratwas committed:
filter empty records

dratwas committed:
(Andorid): Hide keyboard before back, change icon color, remove top corners when full screen

dratwas committed:
hide keyboard first on Android

dratwas committed:
fix icon colors in cell.native

dratwas committed:
Android fixes for animation

dratwas committed:
change the icon color

dratwas committed:
remove old tests

dratwas committed:
remove unused file and some cleanup

dratwas committed:
Merge branch 'master' into rnmobile/feature/link-picker-via-fetch-without-navigation

dratwas committed:
iOS close to cancel / Android arrow left icon to cancel icon

dratwas committed:
Fix initial selection

dratwas committed:
change color of clear icon in link picker

geriux reviewed (commented):
Looking good! Nice work to all involved in this feature!

Just left a small comment.

I do have one question regarding the behavior on an iPad. Should the bottom sheet reach the top? or have some sort of max height?

<kbd><img width="446" alt="Screenshot 2020-09-15 at 15 42 06" src="https://user-images.githubusercontent.com/4885740/93218384-733a0e00-f76a-11ea-9203-efee5f0c11da.png"></kbd>

cc @iamthomasbishop 

Edit: Well actually I just noticed I had the keyboard hidden so when it's visible it doesn't look that big of a space =D

iamthomasbishop commented:
@geriux great catch! I had missed this because I mostly use landscape when on iPad (usually w/ a hardware keyboard) so this is a good reminder to test in portrait more often 😄 ). I agree the height is a bit awkward, but I don't think it's necessarily just a matter of applying a max-height (although that would probably help). I think we should attack this across the board, for tall sheets like in the editor, not necessarily in this specific work (unless y'all think it's an easy-enough fix to include here, of course). In terms of solution, I think we should do something similar to iOS' standard modal sheets, which use following general rules (on tablet):

* Max-height (probably 100% minus some margins on top/bottom, to account for StatusBar and HomeIndicator, other safe areas, provide breathing room, etc.)
* Sheet is centered vertically
* Content container is scrollable

Examples:

| **Landscape** | **Portrait** |
| --------------- | --------------- |
| <a href="https://user-images.githubusercontent.com/1200432/93224836-2d178700-f737-11ea-802e-eb476ec3bad6.jpeg"><img src="https://user-images.githubusercontent.com/1200432/93224836-2d178700-f737-11ea-802e-eb476ec3bad6.jpeg" width="600"></a> | <a href="https://user-images.githubusercontent.com/1200432/93224901-3f91c080-f737-11ea-94b1-9ec72d87b57a.jpeg"><img src="https://user-images.githubusercontent.com/1200432/93224901-3f91c080-f737-11ea-94b1-9ec72d87b57a.jpeg" width="400"></a> |

lukewalczak committed:
Back comment about supported endpoints

lukewalczak commented on packages/react-native-editor/src/api-fetch-setup.js:
```diff
@@ -36,10 +41,7 @@ const fetchHandler = ( { path }, retries = 20, retryCount = 1 ) => {
-		// Please add only wp.org API paths here!
-		/wp\/v2\/(media|categories)\/?\d*?.*/i,
-	].some( ( pattern ) => pattern.test( path ) );
+	SUPPORTED_ENDPOINTS.some( ( pattern ) => pattern.test( path ) );
```
I've added comment back but placed it above the `SUPPORTED_ENDPOINTS`.

geriux replied:
Thank you!

geriux commented:
> I think we should attack this across the board, for tall sheets like in the editor, not necessarily in this specific work (unless y'all think it's an easy-enough fix to include here, of course)

Hey @iamthomasbishop! Thanks for giving examples of a better UI using modal sheets for iPad! Since that involves a bit more work/investigation and it's not entirely related to this PR let's work on it as a separate task and move forward as it is now.

geriux reviewed (approved):
LGTM! Awesome work everyone! 👏

lukewalczak committed:
Merge branch 'master' into rnmobile/feature/link-picker-via-fetch-without-navigation

lukewalczak commented:
> I think we should attack this across the board, for tall sheets like in the editor, not necessarily in this specific work (unless y'all think it's an easy-enough fix to include here, of course)

Thanks for the suggestion! I've created an [issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/2624) to track it.

lukewalczak committed:
Update changelog

lukewalczak committed:
Merge branch 'master' into rnmobile/feature/link-picker-via-fetch-without-navigation

lukewalczak committed:
Add missing colors for link picker

ecgan commented on packages/components/src/mobile/link-picker/link-picker-results.native.js:
```diff
@@ -0,0 +1,125 @@
+		return {
+			fetchMoreSuggestions: debounce( fetchMore, REQUEST_DEBOUNCE_DELAY ),
+		};
+	}, [] );
```
`hasAllSuggestions` should be in the `useSelect` dependency array.

mkevins replied:
Thanks for the comment ecgan :+1: 

It's been a while since I've worked with this code / codebase - but I wonder if simply adding `hasAllSuggestions` to the array might introduce some kind of regression :thinking: . Since `fetchMoreSuggestions` is used in the reset `useEffect`, and is currently expected to be stable, it may need to be added there as well - though I'm not entirely sure off-hand if that can lead to any issues, since the function is debounced (e.g. will this break debouncing in some cases?)

If this is to satisfy a new lint error, and is blocking that work, maybe the simplest solution is adding an ignore - otherwise, this would need further testing in a mobile build to ensure it doesn't introduce regressions, imo.

ecgan replied:
@mkevins , thanks for getting back to me. 🙏 

> I wonder if simply adding `hasAllSuggestions` to the array might introduce some kind of regression 🤔 .

Yes, that's the same concern I have too.

> maybe the simplest solution is adding an ignore

Yeah, I think we can do that. I have just done that in https://github.com/WordPress/gutenberg/pull/49900#discussion_r1171495045.

ecgan reviewed (commented):
👋 Hi! Via https://github.com/WordPress/gutenberg/pull/49900#issuecomment-1513503813, I found that there is a missing dependency in a `useSelect` usage in this PR. I'm not a mobile developer, I'm thinking maybe one of you is in a better position to help fix and test this issue.
